src/libxively/cbor/xi_cbor_codec_ct.c

```diff
@@ -0,0 +1,344 @@
+{
+    ( void )context;
+
+    // printf( "hello calloc wrapper: %lu, %lu\n", count, size );
+
+    return xi_calloc( count, size );
+}
+
+void xi_cn_free_func_wrapper( void* ptr, void* context )
+{
+    ( void )context;
+
+    xi_free( ptr );
+}
+
+static cn_cbor_context cn_cbor_context_object = {.calloc_func = xi_cn_calloc_func_wrapper,
+                                                 .free_func   = xi_cn_free_func_wrapper,
+                                                 .context     = NULL};
+
+cn_cbor_context* context = &cn_cbor_context_object;
+
+#endif
+
+
+#define XI_CBOR_CODEC_CT_STRING_MSGTYPE "msgtype"
+#define XI_CBOR_CODEC_CT_STRING_MSGVER "msgver"
+#define XI_CBOR_CODEC_CT_STRING_LIST "list"
+#define XI_CBOR_CODEC_CT_STRING_FILE_NAME "N"
+#define XI_CBOR_CODEC_CT_STRING_FILE_REVISION "R"
+#define XI_CBOR_CODEC_CT_STRING_FILE_OPERATION "O"
+#define XI_CBOR_CODEC_CT_STRING_FILE_IMAGESIZE "S"
+#define XI_CBOR_CODEC_CT_STRING_FILE_FINGERPRINT "F"
+
+#define XI_CBOR_CODEC_CT_STRING_FILECHUNK_OFFSET "O"
+#define XI_CBOR_CODEC_CT_STRING_FILECHUNK_LENGTH "L"
+
+void xi_cbor_put_name_and_revision( cn_cbor* cb_map,
+                                    const char* name,
+                                    const char* revision,
+                                    cn_cbor_errback* errp )
+{
+    if ( NULL != name )
+    {
+        cn_cbor_map_put( cb_map,
+                         cn_cbor_string_create(
+                             XI_CBOR_CODEC_CT_STRING_FILE_NAME CBOR_CONTEXT_PARAM, errp ),
+                         cn_cbor_string_create( name CBOR_CONTEXT_PARAM, errp ), errp );
+    }
+
+    if ( NULL != revision )
+    {
+        cn_cbor_map_put(
+            cb_map, cn_cbor_string_create(
+                        XI_CBOR_CODEC_CT_STRING_FILE_REVISION CBOR_CONTEXT_PARAM, errp ),
+            cn_cbor_string_create( revision CBOR_CONTEXT_PARAM, errp ), errp );
+    }
+}
+
+void xi_cbor_codec_ct_encode( const xi_control_message_t* control_message,
+                              uint8_t** out_encoded_allocated_inside,
+                              uint32_t* out_len )
+{
+    cn_cbor_errback err;
+    cn_cbor* cb_map = cn_cbor_map_create( CBOR_CONTEXT_PARAM_COMA & err );
+
+    cn_cbor_map_put(
+        cb_map,
+        cn_cbor_string_create( XI_CBOR_CODEC_CT_STRING_MSGTYPE CBOR_CONTEXT_PARAM, &err ),
+        cn_cbor_int_create( control_message->common.msgtype CBOR_CONTEXT_PARAM, &err ),
+        &err );
+
+    cn_cbor_map_put(
+        cb_map,
+        cn_cbor_string_create( XI_CBOR_CODEC_CT_STRING_MSGVER CBOR_CONTEXT_PARAM, &err ),
+        cn_cbor_int_create( control_message->common.msgver CBOR_CONTEXT_PARAM, &err ),
+        &err );
+
+    switch ( control_message->common.msgtype )
+    {
+        case XI_CONTROL_MESSAGE_DB_FILE_INFO:
+
+            if ( 0 < control_message->file_info.list_len )
+            {
+                cn_cbor* files = cn_cbor_array_create( CBOR_CONTEXT_PARAM_COMA & err );
```

Comment:
hm, unfortunately the code formatter does this because of the macro: CBOR_CONTEXT_PARAM_COMA. The formatter does not expand macros.